### DISCUSSION TITLE
Generating Separated Tiles

### DISCUSSION THREAD

**Discussion by helloimalastair** [onthegomap/planetiler#533] [Q&A]
Hey y'all!

Is there a way to get `planetiler` to generate one file per tile, rather than a single file that contains all tiles? Currently working with a single `.pmtiles` file, but looks like Range requests aren't the most performant...

**Comment by msbarry** [comment]
Currently to do this you could generate an mbtiles archive then use a tool like tilelive-copy to extract it to files on disk.

Now that we have an easy way to plug in more archive types though it wouldn't be much work to add a "file" output type that would do this directly.

**Reply by helloimalastair** [reply]
Just took a look at [what appears to be `tilelive-copy`](https://github.com/mapbox/tilelive/blob/master/bin/tilelive-copy), but it looks like it still copies from file to file. Or did you mean rewrite the code myself to manually create each individual tile?

On the topic of the custom archive type, is that a flag I can enable somewhere?

**Reply by msbarry** [reply]
> Just took a look at [what appears to be tilelive-copy](https://github.com/mapbox/tilelive/blob/master/bin/tilelive-copy), but it looks like it still copies from file to file. Or did you mean rewrite the code myself to manually create each individual tile?

In the past I've gotten `tilelive-copy` to work by doing `npm install -g @mapbox/tilelive @mapbox/mbtiles @mapbox/tilelive-file` then `tilelive-copy input.mbtiles file://./tiles`

> On the topic of the custom archive type, is that a flag I can enable somewhere?

Not yet, I opened #536 to track this.

**Comment by bdon** [comment]
@helloimalastair have you experienced a difference between range request latency and non-range requests? From my experience the latency may be inherent to storage like Cloudflare R2 irrespective of whether the `Range` header is used.

**Reply by helloimalastair** [reply]
Hi @bdon!

I was working with a friend who works at Cloudflare with a speed issue on my map, and they confirmed that the latency of `Range` Requests is higher than it should be at the moment, compared to regular `GET` requests, which is something the team wants to improve in the future.

**Reply by bdon** [reply]
@helloimalastair interesting, were you are they able to measure any comparisons in milliseconds between these, and how these correlate to total object size?

**Reply by helloimalastair** [reply]
I don't have any concrete measurements at the moment, though performance appears to get worse(at least on `Range` requests) when the object grows over a few 10s of GB in size.

**Reply by msbarry** [reply]
That would explain some of the performance issues I was seeing trying to serve vector from a pmtiles archive on R2 - each range request was in the 1-2 second range which pretty much made it unusable with up to 3 round trips per tile.

**Reply by bdon** [reply]
See https://github.com/bdon/cloudflare-r2-latency/tree/main for test script on large vs small files and range vs non-range requests. It does seem that range requests for the end of large files are slower than non-range requests for small files, but not by any severe factor (2x, etc), though it is worth investigating more

(the test files were created in the US West region)

**Reply by bdon** [reply]
See https://github.com/protomaps/PMTiles/issues/150#issuecomment-1626470008 for updates on R2

**Reply by msbarry** [reply]
Nice! Do you have any stats on what the new performance looks like?

**Reply by bdon** [reply]
You can try https://github.com/bdon/cloudflare-r2-latency/tree/main - on my EC2 instance near the storage region the results are fairly comparable 180-220ms for all sizes of files now

**Reply by msbarry** [reply]
Ah nice, yeah I'm seeing 350-400ms 
from Boston now.
